Switch to OIDC for SSO (#8414)
cfpb/consumerfinance.gov: docs/sso.md

## Local implementation

Our implementation of OIDC SSO uses a custom subclass of mozilla-django-oidc's `OIDCAuthenticationBackend` that supports additional OIDC claims processing. This enables us to map OIDC roles to Django groups and/or superuser permissioning, and lets us synchronize a user's name to the profile information in the OIDC provider.

This claims processing is implemented via a Django setting that maps an [OIDC claim](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) to a Python callback that will process it. The callable makes what changes it needs to and returns `True` if the user has been modified and should be saved. Our `OIDC_CLAIMS_PROCESSORS` setting looks like this:

```python
OIDC_CLAIMS_PROCESSORS = {
    "roles": "login.auth.process_roles_claim",
    "given_name": "login.auth.process_given_name_claim",
    "family_name": "login.auth.process_family_name_claim",
}
```

Here, `family_name` is processed with `process_family_name_claim`, which is implemented as:

```python
def process_family_name_claim(user, family_name):
    if user.last_name != family_name:
        user.last_name = family_name
        return True
```

> - `OIDC_OP_ADMIN_ROLE`: Django's `is_superuser` will be set for users who are given this role identifier.

Author: Done in [7ef1a61](https://github.com/cfpb/consumerfinance.gov/pull/8414/commits/7ef1a612ad3f0239e320cb6c7a56803230f80a3e)

### Run the test OIDC provider

The test OIDC provider can be run with:

```shell
docker-compose -f docker-compose.sso.yml up
```

This will start the test provider at http://localhost:8080.

### Create a user in the test OIDC provider

To sign-in via the test provider requires a user. mozilla-django-oidc defaults to using email address to map users between the OIDC provider and the client, and we do not override this default.

Navigate to http://localhost:8080 and you will see an option to sign up, as well as log in. Sign up using an email address you know is already set up as a user in your local consumerfinance.gov database.

### Configure the consumerfinance.gov environment

Our `.env_SAMPLE` file includes SSO-related environment variables with defaults intended for local development against the test OIDC provider.

Either uncomment or add the following lines to your `.env` file:

```shell
export ENABLE_SSO=True
export OIDC_RP_CLIENT_ID=4

> export OIDC_RP_CLIENT_SECRET=c99760513bc6

Reviewer: Shouldn't this be changed to match the JSON?

```suggestion
export OIDC_RP_CLIENT_SECRET=itsasecret
```

# Single Sign-On

The consumerfinance.gov codebase support single sign-on (SSO) with [Open ID Connect (OIDC) providers](https://openid.net/). We use Mozilla's [mozilla-django-oidc](https://mozilla-django-oidc.readthedocs.io/en/stable/) library for OIDC support.

## Local implementation

Our implementation of OIDC SSO uses a custom subclass of mozilla-django-oidc's `OIDCAuthenticationBackend` that supports additional OIDC claims processing. This enables us to map OIDC roles to Django groups and/or superuser permissioning, and lets us synchronize a user's name to the profile information in the OIDC provider.

> This claims processing is implemented via a Django setting that maps an [ODIC claim](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) to a Python callback that will process it. The callable makes what changes it needs to and return `True` if the user has been modified and should be saved. Our `OIDC_CLAIMS_PROCESSORS` setting looks like this:

Reviewer: ```suggestion
This claims processing is implemented via a Django setting that maps an [OIDC claim](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) to a Python callback that will process it. The callable makes what changes it needs to and returns `True` if the user has been modified and should be saved. Our `OIDC_CLAIMS_PROCESSORS` setting looks like this:
```

```python
OIDC_CLAIMS_PROCESSORS = {
    "roles": "login.auth.process_roles_claim",
    "given_name": "login.auth.process_given_name_claim",
    "family_name": "login.auth.process_family_name_claim",
}
```

Here, `family_name` is processed with `process_family_name_claim`, which is implemented as:

```python
def process_family_name_claim(user, family_name):
    if user.last_name != family_name:
        user.last_name = family_name
        return True
```

For user roles, we have additional settings. We distinguish two user roles: one for admins, which maps to Django's `is_superuser`, and one for regular users, who are allowed to log into the Wagtail admin. For each there is a setting for the OIDC role:

- `OIDC_OP_ADMIN_ROLE`
- `OIDC_OP_USER_ROLE`

And there is a setting for the name of the default group that all `OIDC_OP_USER_ROLE` users should belong to:

- `SSO_DEFAULT_GROUP`

> If this group does not exist, it will be created. If it does include the `wagtailadmin.access_admin`, that permission will be added.

Reviewer: ```suggestion
If this group does not exist, it will be created. If the group already exists but does not include the `wagtailadmin.access_admin` permission, that permission will be added.
```

## Testing OIDC single sign-on locally

For development of mozilla-django-oidc, Mozilla [uses a test OIDC provider](https://github.com/mozilla/docker-test-mozilla-django-oidc), which is published to Docker Hub as `mozilla/oidc-testprovider:oidc_testprovider-latest`. This test provider is a Django project that uses [django-oidc-provider](https://django-oidc-provider.readthedocs.io/) as the OIDC provider application.

We make use of this test provider to enable local testing and development related to SSO for consumerfinance.gov. The OIDC test provider is available in the `docker-compose.sso.yml` file.

For consumerfinance.gov, our compose file adds another set of fixture data that adds `localhost:8000` as a client in addition to the test provider's built-in [`testrp` client that Mozilla uses for local development](https://github.com/mozilla/mozilla-django-oidc?tab=readme-ov-file#local-development).

> 
> !!! note "Limitations"
> 
>     Mozilla's test provider does not provide group/role claims, just user authentication. This means that, while it enables testing the general SSO flow locally, it does not support provisioning new users in Wagtail (they will not receive access to the Wagtail admin).
> 

Reviewer: ```suggestion
```

Since you changed the logic to remove the check for the regular user role, does this limitation still exist? I created a new user in the test provider and when I logged in, a Wagtail user was created who was able to access the admin.

Author: Removed in 8543484a3
